File: sdk/python/worker/relaymesh/codec.py
import json
from typing import Any, Dict, Optional

from cloud.v1 import githooks_pb2

from .event import Event
from .metadata import (
    METADATA_KEY_EVENT,
    METADATA_KEY_INSTALLATION_ID,
    METADATA_KEY_LOG_ID,
    METADATA_KEY_PROVIDER,
    METADATA_KEY_REQUEST_ID,
)
from .types import RelaybusMessage


class Codec:
    def decode(self, topic: Optional[str], msg: RelaybusMessage) -> Event:
        raise NotImplementedError


class DefaultCodec(Codec):
    def decode(self, topic: Optional[str], msg: RelaybusMessage) -> Event:
        if msg is None or msg.payload is None:
            raise ValueError("message payload is required")

        provider = ""
        event_name = ""
        raw_payload = msg.payload
        normalized: Optional[Dict[str, Any]] = None

        try:
            env = githooks_pb2.EventPayload()
            env.ParseFromString(msg.payload)
            provider = env.provider
            event_name = env.name
            raw_payload = env.payload or b""
            normalized = _parse_json_object(raw_payload)
        except Exception:
            legacy = _parse_json_value(msg.payload)
            if isinstance(legacy, dict):
                provider = str(legacy.get("provider", provider) or provider)
                event_name = str(legacy.get("name", event_name) or event_name)
                data = legacy.get("data")
                if isinstance(data, dict):
                    normalized = data
            if normalized is None:
                normalized = _parse_json_object(raw_payload)

        metadata = dict(msg.metadata or {})
        if not provider:
            provider = metadata.get(METADATA_KEY_PROVIDER, "")
        if not event_name:
            event_name = metadata.get(METADATA_KEY_EVENT, "")

        return Event(
            provider=provider,
            type=event_name,
            topic=_resolve_topic(topic, msg),
            metadata=metadata,
            payload=raw_payload,
            normalized=normalized,
            request_id=metadata.get(METADATA_KEY_REQUEST_ID, ""),
            installation_id=metadata.get(METADATA_KEY_INSTALLATION_ID, ""),
            log_id=metadata.get(METADATA_KEY_LOG_ID, ""),
        )


def _resolve_topic(topic: Optional[str], msg: RelaybusMessage) -> str:
    trimmed = (topic or "").strip()
    if trimmed:
        return trimmed
    return str(msg.topic or "")


def _parse_json_object(data: bytes) -> Optional[Dict[str, Any]]:
    value = _parse_json_value(data)
    if isinstance(value, dict):
        return value
    return None


def _parse_json_value(data: bytes) -> Optional[Any]:
    if not data:
        return None
    try:
        return json.loads(data.decode("utf-8"))
    except Exception:
        return None

File: sdk/go/worker/codec.go
package worker

import (
	"encoding/json"
	"errors"
	"strings"

	relaymessage "github.com/relaymesh/relaybus/sdk/core/go/message"
	"google.golang.org/protobuf/proto"

	cloudv1 "github.com/relaymesh/relaymesh/pkg/gen/cloud/v1"
)

// Codec is an interface for decoding messages from a message broker into an Event.
type Codec interface {
	// Decode transforms a Relaybus message into an Event.
	Decode(topic string, msg *relaymessage.Message) (*Event, error)
}

// DefaultCodec is the default implementation of the Codec interface.
// It decodes a protobuf EventPayload into an Event, with a JSON fallback.
type DefaultCodec struct{}

// Decode unmarshals a Relaybus message into an Event.
func (DefaultCodec) Decode(topic string, msg *relaymessage.Message) (*Event, error) {
	if msg == nil {
		return nil, errors.New("message is nil")
	}
	var (
		provider   string
		eventName  string
		raw        []byte
		normalized map[string]interface{}
	)

	var env cloudv1.EventPayload
	if protoErr := proto.Unmarshal(msg.Payload, &env); protoErr == nil {
		provider = env.GetProvider()
		eventName = env.GetName()
		raw = env.GetPayload()
		if len(raw) > 0 {
			_ = json.Unmarshal(raw, &normalized)
		}
	} else {
		var legacy map[string]interface{}
		if jsonErr := json.Unmarshal(msg.Payload, &legacy); jsonErr != nil {
			return nil, protoErr
		}
		if val, ok := legacy["provider"].(string); ok {
			provider = val
		}
		if val, ok := legacy["name"].(string); ok {
			eventName = val
		}
		if data, ok := legacy["data"].(map[string]interface{}); ok {
			normalized = data
		}
		raw = msg.Payload
	}

	metadata := make(map[string]string, len(msg.Metadata))
	for key, value := range msg.Metadata {
		metadata[key] = value
	}

	if provider == "" {
		provider = metadataValue(msg.Metadata, MetadataKeyProvider)
	}
	if eventName == "" {
		eventName = metadataValue(msg.Metadata, MetadataKeyEvent)
	}

	if normalized == nil {
		var rawJSON interface{}
		if len(raw) > 0 {
			if err := json.Unmarshal(raw, &rawJSON); err == nil {
				if object, ok := rawJSON.(map[string]interface{}); ok {
					normalized = object
				}
			}
		}
	}

	payload := json.RawMessage(raw)
	return &Event{
		Provider:   provider,
		Type:       eventName,
		Topic:      resolveTopic(topic, msg),
		Metadata:   metadata,
		Payload:    payload,
		Normalized: normalized,
	}, nil
}

func metadataValue(meta map[string]string, key string) string {
	if meta == nil {
		return ""
	}
	return meta[key]
}

func resolveTopic(topic string, msg *relaymessage.Message) string {
	if strings.TrimSpace(topic) != "" {
		return topic
	}
	if msg == nil {
		return ""
	}
	return msg.Topic
}
